usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1067 @@
+		if (sep->se_proto != NULL) {
+			if (strncmp(sep->se_proto, "udp", 3) == 0) {
+				sep->se_socktype = SOCK_DGRAM;
+			} else if (strncmp(sep->se_proto, "tcp", 3) == 0) {
+				sep->se_socktype = SOCK_STREAM;
+			}
+		}
+	}
+}
+
+static bool
+setup_internal(struct servtab * sep)
+{
+	pid_t wait_prev = sep->se_wait;
+	if (parse_server(sep, "internal") != 0) {
+		ENI("exec");
+		return false;
+	} else if (wait_prev != SE_WAIT_UNINIT && wait_prev != sep->se_wait) {
+		/* If wait was already specified throw an error. */
+		WRN(WAIT_WRN, sep->se_service, (sep->se_wait ? "yes" : "no"));
+	}
+	return true;
+}
+
+static bool
+infer_protocol_ip_version(struct servtab *sep)
+{
+	struct in_addr tmp;
+
+	if (strcmp("tcp", sep->se_proto) == 0
+		|| strcmp("udp", sep->se_proto) == 0
+		|| strcmp("rpc/tcp", sep->se_proto) == 0
+		|| strcmp("rpc/udp", sep->se_proto) == 0) {
+
+		if (inet_pton(AF_INET6, sep->se_hostaddr, &tmp)) {
+			sep->se_family = AF_INET6;
+		} else if (!inet_pton(AF_INET, sep->se_hostaddr, &tmp)) {
+			ERR("Address family of %s is ambigous or invalid. "
+				"Explicitly specify protocol", sep->se_hostaddr);
+			return false;
+		}
+	}
+	
+	return true;
+}
+
+/* 
+ * Skips whitespaces, newline characters, and comments, 
+ * and returns the next token. Returns false and logs error if an EOF is 
+ * encountered. 
+ */
+static bool 
+skip_whitespace(char **cpp) {
+	char *cp = *cpp;
+
+	int line_start = line_number;
+	
+	for (;;) {
+		while (isblank((int)*cp))
+			cp++;
+
+		if (*cp == '\0' || *cp == '#') {  
+			cp = nextline(fconfig);
+
+			/* Should never expect EOF when skipping whitespace */
+			if (cp == NULL) {
+				ERR("Early end of file after line %d", line_start); 
+				return false;
+			}
+
+		} else break;
+	}
+
+	*cpp = cp;
+	return true;
+}
+
+/* Get the key handler function pointer for the given name */
+static key_handler_func 
+get_handler(char * name) {
+	/* Call function to handle option parsing. */
+	for(unsigned int i = 0; i < A_CNT(key_handlers); i++) {
+		if (strcmp(key_handlers[i].name, name) == 0) {
+			return key_handlers[i].handler;
+		}
+	}
+	return NULL;
+}
+
+/* 
+ * Perform an in-place parse of a single-line quoted string 
+ * with escape sequences. Sets *cpp to the position after the quoted characters.
+ * Uses shell-style quote parsing.
+ */
+static bool 
+parse_quotes(char **cpp) {
+
+	char *cp = *cpp;
+	char quote = *cp;
+
+	memmove(cp, cp+1, strlen(cp)); 	
+	
+	while (*cp && quote) {
+		if (*cp == quote) {
+				quote = '\0';
+				memmove(cp, cp+1, strlen(cp)); 
+		} else if (*cp == '\\') {
+			/* start is location of backslash */
+			char* start = cp;
+			int count_read;
+			cp++;
+			switch (*cp) {
+			case 'x':
+				sscanf(cp+1, "%2hhx%n", start, &count_read);
+				if (count_read == 0) {
+					ERR0("Expected hexadecimal value after \\x");
+					return false;
+				}
+				memmove(cp, cp + count_read + 1, strlen(cp));
+				break;
+			case '\\':
```

Comment:
no point in repeating the memmove so many times. It can be done once at the bottom of the switch if you make the hex case to continue instead of break.